server/service/src/sync/central_data_synchroniser.rs

```diff
@@ -60,76 +78,60 @@ impl CentralDataSynchroniser {
                 .get_central_records(cursor, BATCH_SIZE)
                 .await
                 .map_err(|source| CentralSyncError::PullCentralSyncRecordsError { source })?;
-            let central_sync_records = central_sync_batch_records_to_buffer_rows(sync_batch.data)
-                .map_err(|err| {
-                CentralSyncError::PullCentralSyncRecordsError { source: err.into() }
-            })?;
-
-            if central_sync_records.len() == 0 {
-                info!("Central sync buffer is up-to-date");
-                break;
-            }
 
+            let batch_length = sync_batch.data.len();
             info!(
                 "Inserting {} central sync records into central sync buffer",
-                central_sync_records.len()
+                batch_length
             );
 
-            for central_sync_record in central_sync_records {
-                Self::insert_one_and_update_cursor(&connection, &central_sync_record)
-                    .await
+            for sync_record in sync_batch.data {
+                let buffer_row = sync_record.to_sync_buffer().map_err(|e| {
+                    CentralSyncError::PullCentralSyncRecordsError { source: e.into() }
+                })?;
+                Self::insert_one_and_update_cursor(connection, &buffer_row, sync_record.id)
                     .map_err(
                         |source| CentralSyncError::UpdateCentralSyncBufferRecordsError { source },
                     )?;
             }
-            info!("Successfully inserted central sync records into central sync buffer");
-
-            cursor = central_sync_cursor
-                .get_cursor()
-                .map_err(|source| CentralSyncError::GetCentralSyncCursorRecordError { source })?;
 
-            if cursor >= sync_batch.max_cursor - 1 {
-                info!("All central sync records pulled successfully");
+            if batch_length == 0 {
+                info!("Central sync buffer is up-to-date");
                 break;
             }
         }
         Ok(())
     }
 
     /// insert row and update cursor in a single transaction
-    pub async fn insert_one_and_update_cursor(
+    pub fn insert_one_and_update_cursor(
```

Comment:
maybe not worth it but this could be `insert_multiple_and_update_cursor()` to add a whole batch in one call...

Reply:
I kind of like how `insert_one_and_update_cursors` just return single RepositoryError, `insert_multiple_and_update_cursor` would need to return `CentralSyncError`, and it seems easier to see where possible errors originate if the are mapped in pull. happy to change if you thin it's important